README.md
Exchange arbitration data collection worker - Binance
=====================================================

[![build status](https://gitlab.com/x-data/x-financial-arbitration-data-collection-worker-binance/badges/master/build.svg)](https://gitlab.com/x-data/x-financial-arbitration-data-collection-worker-binance/commits/master)
[![coverage](https://gitlab.com/x-data/x-financial-arbitration-data-collection-worker-binance/badges/master/coverage.svg?job=test)](https://gitlab.com/x-data/x-financial-arbitration-data-collection-worker-binance/commits/master)

This worker aims to collect data from the exchange Binance.

**NOTE:** The `price` field should always be in the `quote asset` value

**NOTE:** This project relies on environment variables to work, so make sure to set up everything before running.

## Running the tests

Requirements:

    * Node (tested with 8.9.4)
    * npm (tested with (5.6.0)

Run:

    npm test

Done. It will check lint issues and run the available tests, displaying a coverage status.

## Installing and using

Here you will find two section, running with docker and without it. We definitely recommend the docker usage for it's simplicity and other several benefits

### Running with docker

Requirements:

    * Docker (tested with 17.12.0-ce)
    * Docker Compose (tested with 1.18.0)

Running the project:

    docker-compose up

### Running without docker

Requirements:

    * Node (tested with 8.9.4)
    * npm (tested with (5.6.0)

Installing dependencies:

    npm install

Running the project:

    npm run local

Environment variables available are listed below and you can check it's usage in the `src/config/default.js` file.

| environment variable    | default                                     | required | description |
|-------------------------|---------------------------------------------|----------|-------------|
| EXCHANGE_NAME           | binance                                     | true     | it becomes *required* when the there is no *SYMBOL_DELIMTER* separating the pair of currencies in the exchange |
| SYMBOL_DELIMITER        |                                             | false    | it defines the pair of currencies separator, e.g. BTC-LTC the `-` is the separator |
| API_URL                 | https://api.binance.com/api/v3/ticker/price | true     | the full API URL route/path you can collect the pair prices |
| COLLECT_ONLY_MATCHES    | false                                       | false    | if you wish to collect only symbols that match with other exchanges |
| COLLECT_SPECIFIC_PRICES | true                                        | false    | if you wish to collect only specific prices and not all the prices available in the exchange |
| COLLECT_PAIRS           | ETH-BTC,BCH-BTC                             | false    | the comma separated list of pairs to collect, e.g. "BTC-LTC, ETH-XRP" |
| INTERVAL_IN_SECONDS     | 300                                         | false    | the interval in seconds to update the pairs |
| NODE_CONFIG_DIR         |                                             | true     | set it to `./src/config` |
| RUNNING_MODE            | single-mode                                 | false    | it specifies if the collect data process should run once (`single-mode`) or forever as a `worker` |
| DB_NAME                 | db                                          | true     | the database name |
| DB_USERNAME             | worker                                      | true     | the username that will login into the database |
| DB_PASSWORD             | worker                                      | true     | the database password |
| DB_HOST                 | localhost                                   | true     | the database host |
| DB_PORT                 | 3306                                        | true     | the database port |
| DB_DEBUG                | false                                       | true     | debugging the database |
| DB_LOGGING              | undefined                                   | true     | whether the database should log or not |
| DB_POOL_MAX             | 20                                          | true     | the maximum pool number |
| DB_POOL_MIN             | 5                                           | true     | the minimum pool number |
| DB_POOL_IDLE            | 10000                                       | true     | the number of pools allowed to be idle |
| LOGGING_LEVEL           | debug                                       | true     | the application logging level |


CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Updated
* project name
* migrate to github
* update node/npm version

# Fixed
* audit fix in packages

## [1.1.0] - 2018-03-14
### Added
* option to collect only matching symbols from another exchanges

## [1.0.0] - 2018-03-02
### Added
* blacklist of invalid pairs
* mapped different host port to database
* worker


src/worker.spec.js
const assert = require('chai').assert
const sinon = require('sinon')
const axios = require('axios')
const config = require('config')
const normalize = require('x-cryptocurrencies-normalizr')

const Worker = require('./worker')
const Data = require('./utils/sequelize').Data

describe('worker', () => {
  const stubRequest = sinon.stub(axios, 'get')
  const stubData = sinon.stub(Data, 'create')

  beforeEach(() => {
    stubRequest.reset()
    stubData.reset()
  })

  after(() => {
    stubRequest.restore()
    stubData.restore()
  })

  it('get data', (done) => {
    const normalizedSymbol = 'BTC-BCH'
    const denormalizedSymbol = 'BCCBTC'

    stubRequest.resolves({data: {price: '1.01230', symbol: denormalizedSymbol}})

    Worker.getData(normalizedSymbol).then((response) => {
      assert.strictEqual(stubRequest.calledOnce, true, 'request was called once')
      assert.strictEqual(
        stubRequest.calledWith(`${config.api.url}?symbol=${denormalizedSymbol}`),
        true,
        'called with API URL and symbol'
      )

      done()
    })
  })

  it('start worker in single-mode', (done) => {
    const pairs = config.collect.pairs.split(',')
    let validPairs = []

    for (let i = 0; i < pairs.length; i++) {
      let normalizedPair = pairs[i]
      let exchangePair = normalize.denormalize.pair(normalizedPair, config.get('exchange').name)

      if (exchangePair !== normalizedPair) {
        validPairs.push(exchangePair)
      }
    }

    stubRequest.resolves({data: {price: '1.01230', symbol: validPairs[0]}})

    Worker.start().then(() => {
      assert.strictEqual(stubData.callCount, validPairs.length, 'saved data for each pair')

      assert.strictEqual(stubRequest.callCount, validPairs.length, 'request was called for each pair')
      assert.strictEqual(
        stubRequest.firstCall.calledWith(`${config.api.url}?symbol=${validPairs[0]}`),
        true,
        'called with API URL and symbol'
      )

      done()
    })
  })
})


src/worker.js
const R = require('ramda')
const axios = require('axios')
const config = require('config')
const normalize = require('x-cryptocurrencies-normalizr')
const logger = require('./utils/logging').logger
const Data = require('./utils/sequelize').Data

const Worker = () => {}

Worker.blackList = []

Worker.getData = async (normalizedPair) => {
  /*
   * The `normalizedPair` parameter is normalized accross all workers
   * since we can't send the request with that value
   * we must first denormalize it
  * */

  let url = config.api.url
  let exchangePair
  if (normalizedPair) {
    exchangePair = normalize.denormalize.pair(normalizedPair, config.get('exchange').name)
    url += `?symbol=${exchangePair}`
  }

  if (normalizedPair === exchangePair) {
    // pair does not exist since exchange pair demands changes
    Worker.blackList.push(normalizedPair)
    return undefined
  }

  const normalizedPairList = normalizedPair.split('-')

  return axios.get(url)
    .then((res) => {
      return {
        baseAsset: normalizedPairList[0],
        quoteAsset: normalizedPairList[1],
        normalizedPair: normalizedPair,
        denormalizedPair: res.data.symbol,
        price: res.data.price
      }
    })
    .catch((err) => {
      if (err.response.status === 400) {
        Worker.blackList.push(normalizedPair)
      }

      logger.error(
        `error requesting ${url}.
         Status: ${err.response.status}.
         Error: ${JSON.stringify(err.response.data)}`
      )
    })
}

Worker.call = async () => {
  const requestList = []

  if (config.collect.onlyMatches === true) {
    let pairs = normalize.findMatches()

    for (let i = 0; i < pairs.length; i++) {
      let pair = pairs[i]

      if (R.indexOf(pair, Worker.blackList) > -1) {
        continue
      }

      requestList.push(
        Worker.getData(pair)
      )
    }
  } else if (config.collect.specificPrices === false) {
    // collect all prices available
    requestList.push(
      Worker.getData()
    )
  } else {
    // collect only a given set of pairs
    if (config.collect.pairs === undefined) {
      throw new Error(
        `in order to collect specific prices you must set \`COLLECT_PAIRS\` with comma separated pairs.
         e.g.: export COLLECT_PAIRS=BTC-ETH,BTC-IOTA,ETH-IOTA
        `
      )
    }

    const coinPairs = R.split(',', config.collect.pairs)

    for (let i = 0; i < coinPairs.length; i++) {
      let coinPair = coinPairs[i]

      if (R.indexOf(coinPair, Worker.blackList) > -1) {
        continue
      }

      requestList.push(
        Worker.getData(coinPair)
      )
    }
  }

  return Promise.all(requestList)
    .then(async (responses) => {
      for (let i = 0; i < responses.length; i++) {
        let response = responses[i]

        if (!response) {
          continue
        }

        logger.debug(response)
        await Data.create(response)
      }
    })
    .catch((err) => {
      logger.error(err)
    })
}

Worker.start = () => {
  logger.info(`starting with ${config.running.mode} mode and interval set to ${config.interval}`)

  return new Promise((resolve, reject) => {
    if (config.running.mode === 'single-time' || config.interval === undefined) {
      Worker.call()
        .then(() => resolve())
        .catch((err) => {
          logger.error(err)
          return reject(err)
        })
      return
    }

    setInterval(Worker.call, config.interval)
    return resolve()
  })
}

module.exports = Worker


src/utils/sequelize.js
const R = require('ramda')
const config = require('config')
const fs = require('fs')
const path = require('path')
const Sequelize = require('sequelize')

const db = config.db
const models = {}

const defaults = {
  logging: config.logging.level === 'debug' ? console.log : undefined,
  operatorsAliases: Sequelize.op,
  define: {
    underscored: true,
    timestamps: false,
    paranoid: false
  }
}

const sequelize = new Sequelize(db.name, db.username, db.password, R.merge(db.options, defaults))

let resourcesPath = path.join(__dirname, '../schemas')
fs.readdirSync(resourcesPath).forEach((resourceName) => {
  let specificResourcePath = path.join(resourcesPath, resourceName)

  fs.readdirSync(specificResourcePath).filter((file) => {
    return (file.indexOf('.schema') >= 0 && file.indexOf('.swp') < 0)
  }).forEach((file) => {
    var model = sequelize.import(file, require(path.join(specificResourcePath, file)))
    var modelsImported = R.type(model) === 'Array' ? model : [model]

    modelsImported.forEach((modelImported) => {
      models[modelImported.name] = modelImported
    })
  })
})

Object.keys(models).forEach((modelName) => {
  if ('associate' in models[modelName]) {
    models[modelName].associate(models)
  }
})

models.sequelize = sequelize
models.Sequelize = sequelize

module.exports = models


migrations/20180224093449_create_exchange_table.js
const config = require('config')

exports.up = (knex, Promise) => {
  return knex.schema.createTable(config.exchange.name, (table) => {
    table.increments('id').unsigned().notNullable().primary()

    // base asset / quote asset / timestamp / normalized pair / denormalized pair / price
    table.string('base_asset', 10).notNullable()
    table.string('quote_asset', 10).notNullable()
    table.string('normalized_pair', 20).notNullable()
    table.string('denormalized_pair', 20).notNullable()
    table.string('price', 20).notNullable()

    table.timestamp('created_at').defaultTo(knex.fn.now())

    // indexes
    table.index('created_at', 'index_created_at')
    table.index(['created_at', 'normalized_pair'], 'index_created_at_normalized_pair')

    table.comment(`${config.exchange.name} data collected`)
    table.engine('InnoDB')
  })
}

exports.down = (knex, Promise) => {
  return knex.schema.dropTable(config.exchange.name)
}


src/config/default.js
const path = require('path')
const env = process.env.NODE_ENV || 'development'
const root = path.join(__dirname, '..', '..')

try {
  require('dotenv').config({path: path.join(root, `.env-${env.toLowerCase()}`)})
} catch (err) {
  console.log(err)
}

const base = {
  exchange: {
    name: process.env.EXCHANGE_NAME || 'binance',
    symbol: {
      delimiter: process.env.SYMBOL_DELIMITER
    }
  },
  api: {
    url: process.env.API_URL || 'https://api.binance.com/api/v3/ticker/price'
  },
  collect: {
    specificPrices: process.env.COLLECT_SPECIFIC_PRICES !== 'false',
    onlyMatches: process.env.COLLECT_ONLY_MATCHES === 'true',
    pairs: process.env.COLLECT_PAIRS || 'ETH-BTC,BCH-BTC,BTC-LTC,BTC-ETH,BTC-XRP,BTC-NEO'
  },
  interval: parseFloat(process.env.INTERVAL_IN_SECONDS || 300) * 1000,
  running: {
    mode: process.env.RUNNING_MODE || 'single-time'
  },
  db: {
    knex: {
      dialect: 'mysql2',
      tablename: 'knex_migrations'
    },
    name: process.env.DB_NAME || 'db',
    username: process.env.DB_USERNAME || 'worker',
    password: process.env.DB_PASSWORD || 'worker',
    options: {
      dialect: 'mysql',
      host: process.env.DB_HOST || 'localhost',
      port: process.env.DB_PORT || 3306,
      debug: process.env.DB_DEBUG || false,
      logging: process.env.DB_LOGGING || undefined,
      pool: {
        max: process.env.DB_POOL_MAX || 20,
        min: process.env.DB_POOL_MIN || 5,
        idle: process.env.DB_POOL_IDLE || 10000
      }
    }
  },
  logging: {
    level: process.env.LOGGING_LEVEL || 'debug',
    project: {
      name: 'x-financial-arbitration-data-collection-worker-binance'
    }
  }
}

module.exports = base
